File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        HashMap<Node, Node> nodes = new HashMap<>();
        Node curr = head;

        while (curr != null) {
            while (curr != null) {
                nodes.put(curr, new Node(curr.val));
                curr = curr.next;
            }
        }

        curr = head;
        while (curr != null) {
            nodes.get(curr).next = nodes.get(curr.next);
            nodes.get(curr).random = nodes.get(curr.random);
            curr = curr.next;
        } 

        return nodes.get(head);
        
    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        previousGroupEnd = dummy

        while True:
            # Find the last node in the next group of k
            groupEnd = previousGroupEnd

            for _ in range(k):
                groupEnd = groupEnd.next
                if not groupEnd:
                    return dummy.next

            nextGroupStart = groupEnd.next
            groupStart = previousGroupEnd.next

            # Reverse this group
            prev = nextGroupStart
            curr = groupStart

            while curr != nextGroupStart:
                nextNode = curr.next
                curr.next = prev
                prev = curr
                curr = nextNode

            # Connect previous group to newly reversed group
            previousGroupEnd.next = groupEnd

            # Original groupStart is now the end of the reversed group
            previousGroupEnd = groupStart